hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-core/src/main/java/org/apache/hadoop/mapreduce/lib/output/FileOutputCommitter.java

```diff
@@ -509,6 +559,28 @@ private void renameOrMerge(FileSystem fs, FileStatus from, Path to,
       if (!fs.rename(from.getPath(), to)) {
         throw new IOException("Failed to rename " + from + " to " + to);
       }
+    } else if (algorithmVersion == 3) {
+      boolean rename = false;
+      if (!fs.exists(to)) {
+        commitJobLock.lock();
+        try {
+          if (!fs.exists(to)) { // double check
```

Comment:
Was thinking if we should not have this double check. Reason being, between line 567 and 568, there could be some other parallel operation which can create `to`. So, we cannot really ensure if its going to be `to` will not be there while renaming. Now at the instant of renaming, two things can happen:
1. `to` path is not there
2. `to` path is there and is a directory
3. `to` path is there and is a file

Now, in rename, it will give false if `to` path is there and is a file (ref: https://hadoop.apache.org/docs/stable/hadoop-project-dist/hadoop-common/filesystem/filesystem.html#boolean_rename.28Path_src.2C_Path_d.29:~:text=HDFS%20%3A%20The%20rename%20fails%2C%20no%20exception%20is%20raised.%20Instead%20the%20method%20call%20simply%20returns%20false), or it will give true. Now two things can happen:
1. `to` was not existing before rename -> after rename `to` path will be made.
2. `to` was an existing dir before rename -> after rename `to/from` will get created (ref: https://hadoop.apache.org/docs/stable/hadoop-project-dist/hadoop-common/filesystem/filesystem.html#boolean_rename.28Path_src.2C_Path_d.29:~:text=If%20the%20destination%20exists%20and%20is%20a%20directory%2C%20the%20final%20destination%20of%20the%20rename%20becomes%20the%20destination%20%2B%20the%20filename%20of%20the%20source%20path.).

Proposal is to:
1. let rename happen
2. if rename happens,  dont do anything more. Reason being, either `to` initially was not there or `to` was directory, then rename would have done similar to what is done on  line 578 to 582